packages/svelte-urql/src/common.ts

```diff
@@ -0,0 +1,119 @@
+import type { Readable, Writable } from 'svelte/store';
+import type { OperationResult } from '@urql/core';
+import { pipe, fromValue, concat, map, makeSubject, onPush } from 'wonka';
+import type { Source } from 'wonka';
+
+/**
+ * An OperationResult annotated with convenience properties
+ *
+ * @todo
+ * React, Vue, and Svelte bindings all use `{fetching:boolean}`
+ * - https://formidable.com/open-source/urql/docs/basics/svelte/#run-a-first-query
+ * - https://formidable.com/open-source/urql/docs/basics/vue/#run-a-first-query
+ *
+ * This probably shouldn't be a Svelte-specific interface, however `@urql/core` doesn't appear to define it
+ *
+ * @todo
+ * preferably, this would also have status properties for convenience:
+ *
+ * - `isIdle`: !fetching && !data
+ * - `isFetching`: !!fetching && !data
+ * - `isSuccess`: !error && !!data
+ * - `isError`: !!error
+ */
+export interface AnnotatedOperationResult<Data, Variables>
+  extends OperationResult<Data, Variables> {
+  /** `true`: awaiting fetch response (no error, no data), otherwise `false` */
+  fetching: boolean;
+}
+
+/**
+ * A Svelte ["Readable" store](https://svelte.dev/docs#run-time-svelte-store-readable) containing Urql [operation results](https://formidable.com/open-source/urql/docs/api/core/#operationresult) along with some convenience tools.
+ */
+export type UrqlStore<Data, Variables extends object = {}> = Readable<
+  AnnotatedOperationResult<Data, Variables>
+>;
+
+/**
+ * Make a Wonka source from a Svelte store
+ * @param store$ a Svelte store
+ * @returns a [Wonka source](https://wonka.kitten.sh/api/sources#makesubject) subscribing to the store's changes
+ */
+export function fromStore<T>(store$: Readable<T>) {
+  const subject = makeSubject<T>();
+  const { source, next } = subject;
+  store$.subscribe(value => next(value));
+  return source;
+}
+
+/**
+ * The starting annotated result object (before fetching)
+ */
+export const defaultBaseResult = {
+  fetching: false,
+  stale: false,
+  error: undefined,
+  data: undefined,
+  extensions: undefined,
+};
+
+/**
+ * Wonka fetching process:
+ *
+ * 1. EMIT: set fetching:true
+ * 2. fetch response(s)
+ * 3. EMIT: build a result from response (fetching:false)
+ * 4. mark the process complete (if isComplete is provided)
+ */
+export function fetchProcess<Data, Variables>(
+  executeOperation: Source<OperationResult<Data, Variables>>,
+  baseResult: AnnotatedOperationResult<Data, Variables>,
+  complete?: () => void
+) {
+  // concat emits the items sequentially
+  return concat([
+    // start with fetching:true
+    fromValue({ ...baseResult, fetching: true }),
+
+    // react to response
+    pipe(
+      // fetch
+      executeOperation,
+
+      // build annotated result from base (set fetching:false)
+      map(response => ({ ...baseResult, ...response })),
+
+      // mark the fetch as complete
+      onPush(() => complete && complete())
```

Comment:
Wonka is not very clear to me, so I would especially appreciate any feedback here.

for example: if the client has a cached copy, i assume it would be emitted from the executeOperation? 

if yes: can anyone suggest a way to specify fetching ONLY during a network request?
if no: how does cache emit?

Reply:
[a comment](https://github.com/FormidableLabs/urql/issues/2371#issuecomment-1085515914) that appears to affirm this implementation:

> as long as the exchange chain is busy we are in theory fetching and that is the only thing a binding should care about.

based on that, it appears we shouldn't try to parse whether there is a network request or not.  fetching is simply a status of whether we're waiting on a response from the client/exchange.

Reply:
That's right, yes. Results basically have a `stale` flag that indicate background fetches / requests. These aren't necessarily tied to an HTTP request happening, and that request isn't assumed to be happening right now. In other words, we assume that there's no need for the UI to know why it's waiting for a result because in most cases we assume it's better to deliver results as soon as possible by default.

That plays into how Wonka's sources work and why they're built synchronous and with predictable ordering in mind. We basically try to impose these requirements on bindings for ourselves:
- When a query is initiated, we execute an operation
- If we get a result immediately then we must never see `fetching: true` as a result (not even as a temporary intermediary step)
- If we don't get a result immediately then we must see `fetching: true` first
- If the operation is changed, e.g. variables change, if the framework for these bindings allow for this, we attempt to keep the previous result unchanged and simply set `fetching: true` again

Reply:
thank you!